File: pytorch/player.py
from abc import ABC, abstractmethod
import numpy as np
import scipy.special
import scipy.stats
from scipy.signal import convolve2d


import env
import network

class Player(ABC):
    def __init__(self) -> None:
        super().__init__()
    
    @abstractmethod
    def play(self, e:env.Env) -> tuple[env.Observation, np.ndarray, env.Action, env.Reward]:
        pass

    @abstractmethod
    def name(self) -> str:
        pass


class ActorPlayer(Player):
    def __init__(self, actor:network.Actor, critic:network.Critic, epoch:int, player:env.Actor) -> None:
        self.actor = actor
        self.critic = critic
        self.player = player
        self.epoch = epoch

    def play(self, e:env.Env) -> tuple[env.Observation, np.ndarray, env.Action, env.Reward]:
        obs = e.observe(self.player)

        device = network.deviceof(self.actor)

        action_probs = self.actor.forward(network.obs_to_tensor(obs, device))[0].to("cpu").detach().numpy()
                
        action_entropy = scipy.stats.entropy(action_probs)
        if action_entropy < 0.001:
            raise ValueError("Entropy is too low!")
                    
        if np.isnan(action_probs).any():
            raise ValueError("NaN found!")
    
        legal_mask = e.legal_mask() 
    
        action_logprobs = np.log(action_probs)
    
        # apply noise to probs
        noise = 0.1*np.random.gumbel(size=len(action_logprobs))
        adjusted_action_probs = scipy.special.softmax(action_logprobs + noise) 
    
        legal_mask = e.legal_mask() 
    
        chosen_action: env.Action = np.argmax(adjusted_action_probs*legal_mask)
        reward,_ = e.step(chosen_action, self.player)
    
        return (
            obs,
            action_probs,
            chosen_action,
            reward
        )
    
    def name(self) -> str:
        return f"actor_ckpt_{self.epoch}"

class RandomPlayer(Player):
    def __init__(self, player:env.Actor) -> None:
        self.player = player
    
    def play(self, e:env.Env) -> tuple[env.Observation, np.ndarray, env.Action, env.Reward]:
        obs = e.observe(self.player)
        legal_mask = e.legal_mask()
        action_prob = scipy.special.softmax(np.random.random(size=len(legal_mask)))
        chosen_action: env.Action = np.argmax(action_prob*legal_mask)
        reward,_ = e.step(chosen_action, self.player)
    
        return (
            obs,
            action_prob,
            chosen_action,
            reward
        )
    
    def name(self) -> str:
        return "random"


# use a convolve2d heuristic to evaluate the board
def heuristic(e:env.Env, player:env.Actor) -> env.Reward:
    # get the opponent
    opponent = env.opponent(player)

    # the board as seen by the player
    player_obs = e.observe(player).board == player
    opponent_obs = e.observe(opponent).board == opponent

    player_score = 0
    opponent_score = 0

    for kernel in env.detection_kernels:
        player_score += np.sum(convolve2d(player_obs, kernel, mode="valid"))
        opponent_score += np.sum(convolve2d(opponent_obs, kernel, mode="valid"))

    score_diff = player_score - opponent_score 
    # logit transform
    return env.Reward(np.exp(score_diff)/(1+np.exp(score_diff)))


# use the minimax algorithm to find the best move, searching up to depth
def minimax(e:env.Env, depth:int, player:env.Actor) -> tuple[env.Reward, env.Action]:
    if depth == 0:
        return (heuristic(e, player), env.Action(0))

    legal_mask = e.legal_mask()
    legal_actions = np.where(legal_mask == 1)[0]

    if len(legal_actions) == 0:
        return (env.Reward(0),env.Action(0))

    best_action = legal_actions[0]
    best_reward:env.Reward = env.Reward(-np.inf)
    for action in legal_actions:
        e.step(action, player)
        reward,_ = minimax(e, depth-1, env.opponent(player))
        e.undo()
        if reward > best_reward:
            best_reward = reward
            best_action = action

    return (best_reward, best_action)

class MinimaxPlayer(Player):
    def __init__(self, player:env.Actor, depth:int) -> None:
        self.player = player
        self.depth = depth
    
    def play(self, e:env.Env) -> tuple[env.Observation, np.ndarray, env.Action, env.Reward]:
        # introduce some randomness
        if np.random.random() > 0.5:
            return RandomPlayer(self.player).play(e)
        
        obs = e.observe(self.player)
        _,chosen_action = minimax(e, self.depth, self.player)
        action_prob = np.eye(e.dims()[1])[chosen_action]
        reward,_ = e.step(chosen_action, self.player)
    
        return (
            obs,
            action_prob,
            chosen_action,
            reward
        )
    
    def name(self) -> str:
        return "minimax"

File: rust/src/player.rs
use crate::env;
use crate::network;
use crate::utils;
use rand::Rng;
use tch::nn::ModuleT;

pub trait Player<const WIDTH: usize, const HEIGHT: usize> {
    /// play one step of the game
    fn play(
        &self,
        s: &mut env::State<WIDTH, HEIGHT>,
    ) -> (
        // the observation
        env::Observation<WIDTH, HEIGHT>,
        // the probability all actions in the action space
        tch::Tensor,
        // the chosen action
        env::Action,
        // the reward for the action
        env::Reward,
    );

    fn name(&self) -> String;
}

pub struct RandomPlayer {
    player: env::Player,
}

impl<const WIDTH: usize, const HEIGHT: usize> Player<WIDTH, HEIGHT> for RandomPlayer {
    fn play(
        &self,
        s: &mut env::State<WIDTH, HEIGHT>,
    ) -> (
        env::Observation<WIDTH, HEIGHT>,
        tch::Tensor,
        env::Action,
        env::Reward,
    ) {
        let legal_moves = s.legal_moves();
        let action = legal_moves[rand::thread_rng().gen_range(0..=legal_moves.len())];
        let (observation, reward) = s.step(self.player, action);
        let prob = tch::Tensor::zeros(&[WIDTH as i64], (tch::Kind::Float, tch::Device::Cpu));
        (observation, prob, action, reward)
    }

    fn name(&self) -> String {
        "RandomPlayer".to_string()
    }
}

pub struct MinimaxPlayer<const WIDTH: usize, const HEIGHT: usize> {
    player: env::Player,
    randomness: f64,
    depth: i64,
}

impl<const WIDTH: usize, const HEIGHT: usize> MinimaxPlayer<WIDTH, HEIGHT>
where
    [(); WIDTH - 4 + 1]:,
    [(); HEIGHT - 4 + 1]:,
    [(); WIDTH - 1 + 1]:,
    [(); HEIGHT - 1 + 1]:,
{
    pub fn new(player: env::Player, depth: i64, randomness: f64) -> Self {
        Self {
            player,
            randomness,
            depth,
        }
    }

    /// use the minimax algorithm to find the best move, searching up to depth
    fn minimax(
        s: &mut env::State<WIDTH, HEIGHT>,
        depth: i64,
        player: env::Player,
    ) -> (env::Reward, usize) {
        if depth == 0 {
            return (Self::heuristic(s.observe(player)), 0);
        }

        let legal_actions = s.legal_moves();

        if legal_actions.len() == 0 {
            return (0.0, 0);
        }

        let mut best_action = legal_actions[0];
        let mut best_reward = f32::NEG_INFINITY;
        for action in legal_actions {
            s.step(player, action);
            let (reward, _) = Self::minimax(s, depth - 1, env::opponent(player));
            s.undo();
            if reward > best_reward {
                best_reward = reward;
                best_action = action;
            }
        }

        (best_reward, best_action)
    }

    const FILTER_HORIZ: [[f32; 4]; 1] = [[1.0, 1.0, 1.0, 1.0]];
    const FILTER_VERT: [[f32; 1]; 4] = [[1.0], [1.0], [1.0], [1.0]];
    const FILTER_DIAG1: [[f32; 4]; 4] = [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];
    const FILTER_DIAG2: [[f32; 4]; 4] = [
        [0.0, 0.0, 0.0, 1.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [1.0, 0.0, 0.0, 0.0],
    ];

    /// check the horizontals, verticals, and diagonals of a boardslice of an observation to calculate a rough heuristic
    fn heuristic_slice(boardslice: &[[bool; WIDTH]; HEIGHT]) -> f32 {
        // cast boardslice to f32 array
        let boardslice = utils::cast2d(boardslice);

        let horiz_out = utils::crosscorrelate2d(&boardslice, &Self::FILTER_HORIZ);
        let vert_out = utils::crosscorrelate2d(&boardslice, &Self::FILTER_VERT);
        let diag1_out = utils::crosscorrelate2d(&boardslice, &Self::FILTER_DIAG1);
        let diag2_out = utils::crosscorrelate2d(&boardslice, &Self::FILTER_DIAG2);

        return utils::sum(&horiz_out)
            + utils::sum(&vert_out)
            + utils::sum(&diag1_out)
            + utils::sum(&diag2_out);
    }

    fn heuristic(obs: env::Observation<WIDTH, HEIGHT>) -> env::Reward {
        let score_diff =
            Self::heuristic_slice(&obs.board[0]) - Self::heuristic_slice(&obs.board[1]);
        f32::exp(score_diff) / (1.0 + f32::exp(score_diff))
    }
}

impl<const WIDTH: usize, const HEIGHT: usize> Player<WIDTH, HEIGHT> for MinimaxPlayer<WIDTH, HEIGHT>
where
    [(); WIDTH - 4 + 1]:,
    [(); HEIGHT - 4 + 1]:,
    [(); WIDTH - 1 + 1]:,
    [(); HEIGHT - 1 + 1]:,
{
    fn play(
        &self,
        s: &mut env::State<WIDTH, HEIGHT>,
    ) -> (
        env::Observation<WIDTH, HEIGHT>,
        tch::Tensor,
        env::Action,
        env::Reward,
    ) {
        let (_, action) = Self::minimax(s, self.depth, self.player);
        let (observation, reward) = s.step(self.player, action);
        let prob = tch::Tensor::zeros(&[WIDTH as i64], (tch::Kind::Float, tch::Device::Cpu));
        (observation, prob, action, reward)
    }

    fn name(&self) -> String {
        "MinimaxPlayer".to_string()
    }
}

struct ActorPlayer<const WIDTH: usize, const HEIGHT: usize> {
    actor: network::Actor<WIDTH, HEIGHT>,
    critic: network::Critic<WIDTH, HEIGHT>,
    player: env::Player,
    epoch: i32,
}

impl<const WIDTH: usize, const HEIGHT: usize> ActorPlayer<WIDTH, HEIGHT> {
    fn new(
        actor: network::Actor<WIDTH, HEIGHT>,
        critic: network::Critic<WIDTH, HEIGHT>,
        epoch: i32,
        player: env::Player,
    ) -> Self {
        Self {
            actor,
            critic,
            player,
            epoch,
        }
    }
}

impl<const WIDTH: usize, const HEIGHT: usize> Player<WIDTH, HEIGHT> for ActorPlayer<WIDTH, HEIGHT>
where
    [(); WIDTH - 4 + 1]:,
    [(); HEIGHT - 4 + 1]:,
    [(); WIDTH - 1 + 1]:,
    [(); HEIGHT - 1 + 1]:,
{
    fn play(
        &self,
        s: &mut env::State<WIDTH, HEIGHT>,
    ) -> (
        env::Observation<WIDTH, HEIGHT>,
        tch::Tensor,
        env::Action,
        env::Reward,
    ) {
        let obs = s.observe(self.player);

        let action_probs = self
            .actor
            .forward_t(&network::obs_to_tensor(&obs, self.actor.device()), false)
            .detach()
            .to_device(tch::Device::Cpu);

        let action_entropy = utils::entropy(&action_probs, tch::Kind::Float);
        if action_entropy < 0.001 {
            panic!("Entropy is too low!");
        }

        let legal_mask = s.legal_mask();

        let action_logprobs = action_probs.log_softmax(0, tch::Kind::Float);

        // apply noise to probs
        let noise = 0.1 * utils::gumbel(&[WIDTH as i64], tch::Kind::Float);
        let adjusted_action_probs = action_logprobs.exp() + noise;

        let chosen_action = i64::from(tch::Tensor::argmax(
            &(adjusted_action_probs * legal_mask),
            0,
            false,
        )) as usize;

        let (observation, reward) = s.step(self.player, chosen_action);

        (observation, action_probs, chosen_action, reward)
    }

    fn name(&self) -> String {
        format!("actor_ckpt_{}", self.epoch)
    }
}
